doc/architecture.md

```diff
@@ -0,0 +1,206 @@
+          `--prefer-s3-redirects`, as these listings are only returned for
+          `GET` requests to a collection, which WebDAV clients don't do.
+          (These listings link blob assets and Zarr entries directly to
+          download URLs rather than `dandidav` resource URLs because doing the
+          latter would result in subsequent requests to `dandidav` that just
+          duplicate work.)
+
+        - Note that this does not apply to download URLs for Zarr entries, as
+          their S3 URLs are assumed to always have the same filename as the
+          entry, and thus signing to add a `Content-Disposition` is
+          unnecessary.
+
+- [`DandiDav::get_resource()`][] and
+  [`DandiDav::get_resource_with_children()`][] perform method calls on a
+  `DandiClient` or `ZarrManClient` instance (as appropriate) belonging to the
+  `DandiDav` instance in order to fetch information about the specified
+  resource(s).
+
+
+`DandiClient`
+-------------
+
+- The [`DandiClient`][] type is used to retrieve information about resources on
+  an Archive instance (the one specified via the `--api-url` option on program
+  invocation) via a combination of the Archive's API and (for entries & folders
+  inside Zarrs) requests to an S3 bucket.  It is the data source for the
+  `/dandisets/` hierarchy served by `dandidav`.
+
+    - Information about child resources of Zarrs needs to be fetched via S3
+      directly rather than using the Archive API because the latter currently
+      does not provide an efficient way to list folders within a Zarr.
+
+- The `DandiClient` API is modeled somewhat after
+  [`octocrab`](https://github.com/XAMPPRocky/octocrab): To operate on an API
+  resource, an endpoint object is requested from the `DandiClient` (possibly
+  indirectly via a super-endpoint), and the endpoint's methods are called to
+  obtain one or more objects that provide information about the resource or its
+  children, but these resource objects are distinct from the endpoint objects
+  and do not have any methods for making API calls themselves.
+
+- The code assumes Zarr assets have a `contentUrl` that points to the Zarr's
+  entries on an S3 bucket; specifically, the first element of `contentUrl` that
+  can be parsed by [`S3Location::parse_url()`][s3loc-parse] into a bucket,
+  optional region, and key prefix is used as such.  It is then assumed that the
+  Zarr's entries are laid out under the given key prefix (after appending a
+  trailing slash if one is not already present) on the given bucket with the
+  same names & directory structure as the actual Zarr.
+
+    - If a given Zarr asset lacks an S3 `contentUrl`, any requests to a path
+      under that Zarr will result in a 500 response.
+
+    - If an S3 URL does not specify the bucket region, it is determined via a
+      `HEAD` request to the domain.  The S3 client caching (see below) means
+      that this will almost certainly be done at most once per `dandidav`
+      process.
```

Comment:
just a note: i don't know much of AWS/S3 and AFAIK not applicable to dandiarchive ATM, but I wonder if there are buckets which have multizone configuration then webdav server might be in a different zone from a client... then ideally determining zone should somehow be done and provided by client as may be webdav option etc.